124-exec-engine.Rmd

```diff
@@ -0,0 +1,190 @@
+---
+title: About `exec` engine
+editor: visual 
+---
+
+# How does this engine works ?
+
+The `exec` engine allows to execute an arbitrary command on the code chunk content, optionally with arguments. If any, they are passed to `engine.opts` and they could be one of the following:
+
+``` r
+engine.opts = list(command, file, ext, clean, args, args1, args2)
+```
+
+The execution of the chunk will be equivalent to this command line
+
+    <command> <args1> <args> <args2>
+
+Behavior by default, during evaluation will be:
+
+-   Retrieve the path of the command in `command` option.
+-   Write the chunk content into a temporary file `f`
+-   Register the `clean` function to execute on exit, basically to remove the temp file `f`.
+-   Compute the argument of the command line by appending the results of `args1()`, with `args()` then `args2()`.
+
+There should be very rare usage of changing the default behavior, but it can be useful to debug for example or other advanced usage.
+
+You can find some examples below on how to use each option. We are using `Rscript` an d `python` only in the examples below by convenience but this work with any executable file.
+
+# Simple example
+
+`command` can be the name of a command in PATH.
+
+```{exec, command='Rscript'}
+1 + 1
+```
+
+```{exec, command='python'}
+print(1 + 1)
+```
+
+or a full path
+
+```{exec, command = Sys.which('python'), message = FALSE}
+print("Full path command works")
+```
+
+It can also be passed as an option in `engine.opts`.
+
+```{exec, engine.opts = list(command ='python')}
+print(1 + 1)
+```
+
+It is the only option that can be passed as first level chunk option.
+
+# Options
+
+All other options should be provided in the chunk option `engine.opts` in a list. (As seen above, `command` can be optionally in `engine.opts`, too).
+
+They can take character values or functions that returns character values.
+
+## file
+
+It returns the temporary file path that will be used for execution by the command.
+
+It is a function of two argument `function(code, file){}` - By default, function is writing `code` into the temporary file `file` that will be executed by the command. This can be change if needed for other side effect. You would still want to write the `code` into a file (using `xfun::write_utf8()` or another function) otherwise the command will be executed with an empty file.
```

Comment:
As I said in the other comment, `file` can only be a function `function(code, file)` which contains a writing function to write `code`  into `file` otherwise, the file that will be passed to the executable will be empty. 

Should we modify the `exec` engine so that `file` option can be used to rename the temp file ?

Reply:
Yes, and done in knitr.